### ISSUE TITLE
Replica with extra_port shows up twice in Orchestrator

### ISSUE THREAD

**Issue by dougfales** [openark/orchestrator#923]
We use `gh-ost` on some of our replicas using the [`extra_port` option](https://www.percona.com/doc/percona-server/LATEST/performance/threadpool.html#extra_port), which means we often have the same replica showing up twice in Orchestrator's topology graph like so:

![ghost_extra_port_duplicate_instance](https://user-images.githubusercontent.com/10288/60590030-2f6f0a80-9d58-11e9-8735-f0cb632f02b5.png)

It would be nice to be able to prevent discovery of nodes on this `extra_port`, so that once they are forgotten, they stay forgotten (and undiscoverable) forever. 

The `DiscoveryIgnoreReplicaHostnameFilters` might work for us if this instance had a different hostname, but it only differs in port, and we do want it to appear in Orchestrator, just only under its standard port.

Assuming I'm not missing a config option that already does this, I'd like to add support for a `DiscoveryIgnorePortFilter` setting similar to the `DiscoveryIgnoreReplicaHostnameFilters` setting. 

Alternatively, we could extend the `DiscoveryIgnoreReplicaHostnameFilters` setting to match against the full `StringCode()` (hostname:port) so that I could use a setting like:

```
  "DiscoveryIgnoreReplicaHostnameFilters": ["node1:16022"]
```

But that would change existing behavior for this setting, which is why I was leaning toward a separate setting for ignoring ports. (Related: I see there is already [some work being done](https://github.com/github/orchestrator/pull/346) on updating the `DiscoveryIgnoreReplicaHostnameFilters` setting. Maybe I will propose this there as well.)

My question: would you consider a PR that supports this port-ignoring behavior? If so, would you prefer a separate a setting or extending the behavior of `DiscoveryIgnoreReplicaHostnameFilters`?

**Comment by shlomi-noach** [comment]
@dougfales I'm not sure this is an `extra_port` issue. I wonder which version of `orchestrator` are you using? I've already fixed the issue of `orchestrator` "identifying" `gh-ost` as if it were a replica. I think this would work regardless of the port.

**Comment by dougfales** [comment]
@shlomi-noach we're using gh-ost 1.0.48,  orchestrator 3.0.14, and percona server 5.7.20. I'm able to reproduce the issue locally by running a test-on-replica migration against a replica using the extra port. 

Should I try with 3.1.0 or was your fix pre-3.0.14?

**Comment by shlomi-noach** [comment]
`3.0.14` should be good. 

I think I might understand now how this happens. Let me look into this

**Comment by shlomi-noach** [comment]
Sorry for the late followup. Am I correct that `extra_port` only exists in Percona Server and MariaDB? I seem to remember there was something similar for MySQL but can't find it right now.

**Comment by zmoazeni** [comment]
@shlomi-noach I believe you're correct. I tried to do some searching on MySQL Enterprise (which also offers a thread pool), but came up with nothing

**Comment by zmoazeni** [comment]
The current relevant code is:

https://github.com/github/orchestrator/blob/f6fc25cfe0aec2f153d07f3f4819f9fae544fad0/go/logic/orchestrator.go#L282-L284

and 

https://github.com/github/orchestrator/blob/f6fc25cfe0aec2f153d07f3f4819f9fae544fad0/go/inst/instance_utils.go#L252-L260

I believe we could extend this without changing exiting behavior by:

1. Including the port along with `replicaKey.Hostname`
2. Changing the regexp filter to end in a wildcard

I may be missing an obvious edge case, but I think that would support existing configurations while simultaneously allowing new values to include the port as part of the match.

**Comment by shlomi-noach** [comment]
@zmoazeni thanks! Not sure I'm following; is your suggestion that the users configure `DiscoveryIgnoreReplicaHostnameFilters` to include all their hostnames with `:<extra_port>`? I was thinking more like double-checking the `port` vs `extra_port` after having been able to connect to a server.

**Comment by zmoazeni** [comment]
> is your suggestion that the users configure DiscoveryIgnoreReplicaHostnameFilters to include all their hostnames with :<extra_port>?

Yeah this is what I was thinking. Rather than having orchestrator deal with `extra_port` logic at all, we can could offer the ability for folks who are operating servers with extra ports to explicitly ignore them.

I guess one caveat (that I haven't tested) is that they could also configure a `DiscoveryIgnoreReplicaHostnameFilters` of "*:3307" (3306 being the regular port and 3307 being the extra) to ignore all extra ports on their topology that run on 3307.

This is wild guess here totally not backed by numbers, but I think the number of people who are using thread pools+extra_port+using orchestrator+migrating gh-ost through it is super super low.

**Comment by shlomi-noach** [comment]
Ah, got it! A `"DiscoveryIgnoreReplicaHostnameFilters" : [ "*:3307" ]`-like solution would be very good indeed!

**Comment by shlomi-noach** [comment]
@dougfales So, as @zmoazeni points out, `DiscoveryIgnoreReplicaHostnameFilters` already supports regular expressions (courtesy @sjmudd). Thus, a solution would be:
```json
  "DiscoveryIgnoreReplicaHostnameFilters": [
    ".*:3307"
  ],
```

assuming your `extra_port` is `3307`.

**Comment by zmoazeni** [comment]
@shlomi-noach almost. I was working with @dougfales at the time we saw this popping up. (Long before he submitted this issue). And I remember trying that and failing.

orchestrator still needs to be modified to include the port because `replicaKey.Hostname` is only the hostname portion (doesn't include `:3307` to match via regexp)

https://github.com/github/orchestrator/blob/f6fc25cfe0aec2f153d07f3f4819f9fae544fad0/go/logic/orchestrator.go#L282-L284

**Comment by shlomi-noach** [comment]
@zmoazeni ahhhhhhh... definitely. `orchestrator` will have to use the full `hostname:port` here.

**Comment by shlomi-noach** [comment]
replace
```
if inst.RegexpMatchPatterns(replicaKey.Hostname, config.Config.DiscoveryIgnoreReplicaHostnameFilters) { 
 	continue 
 } 
```
with
```
if inst.RegexpMatchPatterns(replicaKey.StringCode(), config.Config.DiscoveryIgnoreReplicaHostnameFilters) { 
 	continue 
 } 
```

**Comment by dougfales** [comment]
Thanks @shlomi-noach and @zmoazeni ! That's exactly what I was hoping for because it is the simplest option. I think I tried exactly this in a branch while I was reproducing the issue. I'll see if I can turn it into a PR some time this week.

**Comment by dougfales** [comment]
Ok @shlomi-noach @zmoazeni here's the simple PR: https://github.com/github/orchestrator/pull/952

I also think it would be great if at some point https://github.com/github/orchestrator/pull/346 could be merged because it does the ignoring/filtering at multiple points in the discovery flow. In my testing, I observed that, with the code as it is today, an ignored host could potentially be re-discovered, because the ignoring is done in just one place, and depending on how the hostname resolves in Orchestrator, it could slip through the filter.

To be clear, I'm totally happy with the change we arrived at in this issue and my PR. It will definitely be useful to us. Thanks!
